docs/tutorials/q_learning_torch.py

```diff
@@ -0,0 +1,190 @@
+            batch_first=True
+        ),
+        "rewards": torch.nn.utils.rnn.pad_sequence(
+            [torch.as_tensor(x.rewards) for x in batch],
+            batch_first=True
+        ),
+        "terminations": torch.nn.utils.rnn.pad_sequence(
+            [torch.as_tensor(x.terminations) for x in batch],
+            batch_first=True
+        ),
+        "truncations": torch.nn.utils.rnn.pad_sequence(
+            [torch.as_tensor(x.truncations) for x in batch],
+            batch_first=True
+        )
+    }
+
+# %%%
+# We can now proceed to load the data and create the training loop. To begin, let's initialize the DataLoader, neural network, optimizer, and loss.
+
+minari_dataset =  minari.load_dataset("CartPole-v1-random")
+dataloader = DataLoader(minari_dataset, batch_size=64, shuffle=True, collate_fn=collate_fn)
+
+env = minari_dataset.recover_environment()
+observation_space = env.observation_space
+action_space = env.action_space
+assert isinstance(observation_space, spaces.Box)
+assert isinstance(action_space, spaces.Discrete)
+
+q_net = QNetwork(np.prod(observation_space.shape), action_space.n)
+optimizer = torch.optim.Adam(q_net.parameters(), lr=0.001)
+loss_fn = nn.MSELoss()
+
+# %%%
+# We use the Mean Square Error loss, as a classic regression task. In fact, we want to predict the `Q`-values of each state-action pairs that we take from the trajectories. 
+# The training loop will look like this:
+
+num_epochs = 3
+gamma = 0.99
+
+for epoch in range(num_epochs):
+    for batch in tqdm(dataloader):
+        q_values = q_net(batch['observations'])
+        q_targets = batch['rewards'].float() + gamma * torch.max(q_values[:, 1:], dim=-1).values
+        q_pred = q_values[:, :-1].gather(-1, batch['actions'].unsqueeze(-1)).squeeze()
+            
+        mask = 1 - torch.cumsum(
+            torch.logical_or(batch["terminations"], batch["truncations"]),
+            dim=-1
+        )
+        loss = loss_fn(q_pred * mask, q_targets * mask)
+
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+    print(f"Epoch: {epoch}/{num_epochs}, Loss: {loss.item()}")
+
+# %%%
+# where we compute the ``q_targets`` using the SARSA objective 
+# 
+# .. math ::
+#   r(s_t, a_t) + \gamma \max_{a' \in \mathcal{A}} \hat{Q}_{\theta}(s_{t+1}, a')
+#
+# Similarly, the loss for a single episode will be: 
+#
+# .. math ::
+#   l(\theta) = \sum_{t=0}^{T-1} \Big( r(s_t, a_t) + \gamma \max_{a' \in \mathcal{A}} \hat{Q}_{\theta}(s_{t+1}, a') - \hat{Q}_{\theta}(s_t, a_t) \Big)^2
+# 
+# Where `T` is the toal number of step in the episode. However, we have a batch of episodes with different length `T`; we use ``mask`` to zeros the loss after the episode is terminated. 
+# 
+# Lastly, we can assess the performance of the agent that utilizes the learned `Q`-values:
+
+env = gym.make("CartPole-v1", render_mode="human")
+obs, _ = env.reset(seed=42)
+done = False
+accumulated_rew = 0
+while not done:
```

Comment:
For reference, I think the code is not fully deterministic since I got:
````
Offline Q-Learning accumulated rew:  160.0
Dataset mean accumulated rew:  22.12796
````